### ISSUE TITLE
Overall structure of EpiAware: proposing SciML style

### ISSUE THREAD

**Issue by SamuelBrand1** [CDCgov/Rt-without-renewal#122]
The goal of this issue is to define the overall structure of `EpiAware` API; and to suggest an opinionated approach.

I think we should try and match the `SciML` pattern of:

- [x] #145
- [x] #151
- [x] #156

### EpiAware problem object
Propose calling the type `EpiAwareProblem`.
This should contain:
  - A latent model `<: AbstractLatentModel`
  - An epimodel `<: AbstractEpiModel`
  - An observation model `<: AbstractObservationModel`. _Any actual data we have should be a random variable in the Observation model and we can use variable passing or `DynamicPPL` conditioning_. Discussed this briefly with @seabbs off-line.

### Solution method problem object
Propose calling the type `EpiAwareInferenceMethod`.
This should contain:
- Initialisation method
- sampler type e.g. NUTS

### Our `solve` function
I think we should call this `epi_aware` and have signature

```julia
epi_prob = EpiAwareProblem(...)
inference_method = EpiAwareInferenceMethod(...)
sol = epi_aware(epi_prob, inference_method; kwargs...)
```

**Comment by seabbs** [comment]
> I think we should try and match the SciML pattern of:

Strongly agree.

>  we can use variable passing or DynamicPPL conditioning. 

Agree and vote for conditioning as it seems very nice. Only issue I see is knowing precisely what to condition on and that its working as expected.

> An observation model <: AbstractObservationModel. 

To generalise this do we want hypothetical support for n `AbstractObservationModel`s or would we instead create a nested type to deal with this?

**Comment by SamuelBrand1** [comment]
> To generalise this do we want hypothetical support for n AbstractObservationModels or would we instead create a nested type to deal with this?

My vote would be for a nested type. 

It would be nice to have a Turing (sub)model for observation where you could pass in:
- A times series of true/latent incidence
- parameter priors

And then you could sample observations. Then everything from a trivial "You just observes incidence as it occurs" through to a more complicated delayed obs and reporting model designed for nowcasting could all be accepted.

**Comment by SamuelBrand1** [comment]
Yeah I think that is fine, I don't think we strong type the data anywhere. If we have an observation model that has something like 

`y_t ~ LogNormal(...)` 

in there it _should_ just work

**Comment by seabbs** [comment]
I mean if we have a mix of obs models defined within a hierarchy of obs and if we rely on only y_t as the input to that hierarchy then it will have a mixed type. Perhaps that isn't a problem here?

**Comment by SamuelBrand1** [comment]
> I mean if we have a mix of obs models defined within a hierarchy of obs and if we rely on only y_t as the input to that hierarchy then it will have a mixed type. Perhaps that isn't a problem here?

I envision a hierarchy of obs and needing a super model which
1. Generates hyperparameters from a set of hyperpriors, then parameters from any parameter population distribution
2. These are passed to EpiAware models for each observation set.

e.g. if we have two signals: one is count and one is proportions then there would be some EpiAware model for each source and they live in a wider model which interacts with them via conditioning parameters that come from step 1.

**Comment by SamuelBrand1** [comment]
Otoh I can also imagine doing this but the observations are only based on a single infection time series... interesting problem.

**Comment by seabbs** [comment]
I think this has now been resolved in #169 with a slight deviation of calling our `solve` `apply_method` but functionality wise is the same
